lib/backends/smartos/tasks/recovery_config.js

```diff
@@ -0,0 +1,282 @@
+    var buf = Buffer.from(hexStr, 'hex');
+    // variant:
+    buf[8] = buf[8] & 0x3f | 0xa0;
+    // version:
+    buf[6] = buf[6] & 0x0f | 0x50;
+    var hex = buf.toString('hex', 0, 16);
+    const uuid = [
+        hex.substring(0, 8),
+        hex.substring(8, 12),
+        hex.substring(12, 16),
+        hex.substring(16, 20),
+        hex.substring(20, 32)
+    ].join('-');
+    return uuid;
+}
+
+
+function RecoveryConfigTask(req) {
+    Task.call(this);
+    this.req = req;
+}
+
+var logger = bunyan.createLogger({
+    name: 'cn-agent',
+    stream: process.stderr,
+    level: 'debug',
+    serializers: bunyan.stdSerializers
+});
+
+
+Task.createTask(RecoveryConfigTask);
+
+function start() {
+    var self = this;
+
+    var kbmapiaddr;
+    var server_uuid;
+    var guid = self.req.params.pivtoken;
+    var template = self.req.params.template;
+    var action = self.req.params.action;
+    var recovery_uuid = self.req.params.recovery_uuid;
+    var token = self.req.params.token;
+    var templateFile;
+    var zpoolRecovery = {};
+    var zpool;
+
+    async.waterfall([
+        function retrieveKbmapiAddresses(cb) {
+            smartdc_config.sdcConfig(function (error, cfg) {
+                if (error) {
+                    cb(error);
+                    return;
+                }
+
+                var domainName = 'kbmapi.' + cfg.datacenter_name + '.' +
+                    cfg.dns_domain;
+
+                logger.info({
+                    domainName: domainName
+                }, 'kbmapi domain name');
+
+                dns.resolve(domainName, function (dnserror, addrs) {
+                    if (dnserror) {
+                        cb(dnserror);
+                        return;
+                    }
+
+                    if (!addrs.length) {
+                        cb('No KBMAPI addresses found');
+                        return;
+                    }
+
+                    kbmapiaddr = addrs[0];
+                    self.progress(10);
+                    cb();
+                });
+            });
+        },
+
+        function getSysinfo(cb) {
+            execFile('/usr/bin/sysinfo', ['-f'], function (err, stdo, stde) {
+                if (err) {
+                    cb(Error(stde.toString()));
+                    return;
+                }
+                var obj = JSON.parse(stdo.toString());
+                var zpoolEncryption = obj['Zpool Encrypted'];
+                if (!zpoolEncryption) {
+                    cb(Error('Recovery configuration can be staged or ' +
+                        'activated only on servers with encrypted zpools'));
+                    return;
+                }
+                zpoolRecovery = obj['Zpool Recovery'] || {};
+                zpool = obj['Zpool'];
+                server_uuid = obj.UUID;
+                self.progress(25);
+                cb();
+            });
+        },
+
+        function bailIfMissingPreconditions(cb) {
```

Comment:
I think most of this validation function can/should be first in the vasync.pipeline (i.e. validate first, setup second).

Reply:
Has sense. Done.